var/spack/repos/builtin/packages/gromacs/package.py

```diff
@@ -193,49 +193,36 @@ class Gromacs(CMakePackage, CudaPackage):
-    depends_on("plumed@2.7.1:2.7.2+mpi", when="@2021+plumed+mpi")
-    depends_on("plumed@2.7.1:2.7.2~mpi", when="@2021+plumed~mpi")
-    depends_on("plumed@2.8.1:2.8.2+mpi", when="@2020.7+plumed+mpi")
-    depends_on("plumed@2.8.1:2.8.2~mpi", when="@2020.7+plumed~mpi")
-    depends_on("plumed@2.7.2:2.8+mpi", when="@2020.6+plumed+mpi")
-    depends_on("plumed@2.7.2:2.8~mpi", when="@2020.6+plumed~mpi")
-    depends_on("plumed@2.7.1+mpi", when="@2020.5+plumed+mpi")
-    depends_on("plumed@2.7.1~mpi", when="@2020.5+plumed~mpi")
-    depends_on("plumed@2.6.2:2.7.0+mpi", when="@2020.4+plumed+mpi")
-    depends_on("plumed@2.6.2:2.7.0~mpi", when="@2020.4+plumed~mpi")
-    depends_on("plumed@2.6.1+mpi", when="@2020.2+plumed+mpi")
-    depends_on("plumed@2.6.1~mpi", when="@2020.2+plumed~mpi")
-    depends_on("plumed@2.6.1:2.8.2+mpi", when="@2019.6+plumed+mpi")
-    depends_on("plumed@2.6.1:2.8.2~mpi", when="@2019.6+plumed~mpi")
-    depends_on("plumed@2.5.3:2.6.0+mpi", when="@2019.4+plumed+mpi")
-    depends_on("plumed@2.5.3:2.6.0~mpi", when="@2019.4+plumed~mpi")
-    depends_on("plumed@2.5.2+mpi", when="@2019.2+plumed+mpi")
-    depends_on("plumed@2.5.2~mpi", when="@2019.2+plumed~mpi")
-    depends_on("plumed@2.5.3:2.6+mpi", when="@2018.8+plumed+mpi")
-    depends_on("plumed@2.5.3:2.6~mpi", when="@2018.8+plumed~mpi")
-    depends_on("plumed@2.5.1:2.5.2+mpi", when="@2018.6+plumed+mpi")
-    depends_on("plumed@2.5.1:2.5.2~mpi", when="@2018.6+plumed~mpi")
-    depends_on("plumed@2.5.0+mpi", when="@2018.4+plumed+mpi")
-    depends_on("plumed@2.5.0~mpi", when="@2018.4+plumed~mpi")
-    depends_on("plumed@2.5.1:2.5+mpi", when="@2016.6+plumed+mpi")
-    depends_on("plumed@2.5.1:2.5~mpi", when="@2016.6+plumed~mpi")
-    depends_on("plumed@2.5.0+mpi", when="@2016.5+plumed+mpi")
-    depends_on("plumed@2.5.0~mpi", when="@2016.5+plumed~mpi")
+    with when("+plumed"):
+        plumed_patches = {
+            "2022.5": "2.8.2",
+            "2022.3": "2.8.1",
+            "2021.7": "2.8.2",
+            "2021.6": "2.8.1",
+            "2021.5": "2.7.5:2.7.6",
+            "2021.4": "2.7.3:2.8.0",
+            "2021": "2.7.1:2.7.2",
+            "2020.7": "2.8.1:2.8.2",
+            "2020.6": "2.7.2:2.8.0",
+            "2020.5": "2.7.1",
+            "2020.4": "2.6.2:2.7.0",
+            "2020.2": "2.6.1",
+            "2019.6": "2.6.1:2.8.2",
+            "2019.4": "2.5.3:2.6.0",
+            "2019.2": "2.5.2",
+            "2018.8": "2.5.3:2.6",
+            "2018.6": "2.5.1:2.5.2",
+            "2018.4": "2.5.0",
+            "2016.6": "2.5.1:2.5",
+            "2016.5": "2.5.0",
+        }
+        for gmx_ver, plumed_vers in plumed_patches.items():
+            depends_on(
+                "plumed@{0}+mpi".format(plumed_vers), when="@{0}+plumed+mpi".format(gmx_ver)
+            )
+            depends_on(
+                "plumed@{0}~mpi".format(plumed_vers), when="@{0}+plumed~mpi".format(gmx_ver)
+            )
```

Comment:
What I mean is roughly the following (not tested, so beware of typos):
```suggestion
        for gmx_ver, plumed_vers in plumed_patches.items():
            depends_on(
                "plumed@{0}".format(plumed_vers), when="@{0}".format(gmx_ver)
            )
        depends_on("plumed@~mpi", when="~mpi")
        depends_on("plumed@+mpi", when="+mpi")
```
So  20 (versions) + 2 (mpi variant) conditions instead of 20 * 2

Reply:
I understand that, the issue with that is that there are versions of gromacs that don't have plumed patches yet, and maybe never will. So the user asks for `gromacs@2023 +mpi +plumed` (a version without plumed support) which satisfies `depends_on("plumed+mpi", when="+mpi")` and the build eventually fails. 

So I guess the actual alternatives would be:
- 20 `depends_on` specific version combinations
- 2 `depends_on` generic `+mpi` `~mpi` variants
- `conflict`s for the 26 versions of gromacs that don't have support for plumed, that has to be maintained when new versions of gromacs are added

Versus the explicit list of 20*2 `depends_on` in this PR.

Reply:
Are we saying that right now `gromacs@2023+plumed` is valid but doesn't have plumed?

Reply:
Yes, that's the problem. Conditional variants are a good idea, updating the PR now.